tests/test_app/test_composable.py

```diff
@@ -39,294 +42,383 @@
-        got = proc.apply_to([str(m) for m in dstore], show_progress=False)
-        self.assertEqual(len(got), len(dstore))
-        # or a single string
-        got = proc.apply_to(str(dstore[0]), show_progress=False)
-        self.assertEqual(len(got), 1)
-        self.assertIsInstance(got[0], SequenceCollection)
-        # raises ValueError if empty list
-        with self.assertRaises(ValueError):
-            proc.apply_to([])
-
-        # raises ValueError if list with empty string
-        with self.assertRaises(ValueError):
-            proc.apply_to(["", ""])
-
-    def test_apply_to_strings(self):
-        """apply_to handles strings as paths"""
-        dstore = io_app.get_data_store("data", suffix="fasta", limit=3)
-        dstore = [str(m) for m in dstore]
-        with TemporaryDirectory(dir=".") as dirname:
-            reader = io_app.load_aligned(format="fasta", moltype="dna")
-            min_length = sample_app.min_length(10)
-            outpath = os.path.join(os.getcwd(), dirname, "delme.tinydb")
-            writer = io_app.write_db(outpath)
-            process = reader + min_length + writer
-            # create paths as strings
-            r = process.apply_to(dstore, show_progress=False)
-            self.assertEqual(len(process.data_store.logs), 1)
-            process.data_store.close()
+def test_composable():
+    """correctly form string"""
 
-    def test_apply_to_non_unique_identifiers(self):
-        """should fail if non-unique names"""
-        dstore = [
-            "brca1.bats.fasta",
-            "brca1.apes.fasta",
-        ]
-        with TemporaryDirectory(dir=".") as dirname:
-            reader = io_app.load_aligned(format="fasta", moltype="dna")
-            min_length = sample_app.min_length(10)
-            process = reader + min_length
-            with self.assertRaises(ValueError):
-                process.apply_to(dstore)
+    class app_dummyclass_1:
+        def __init__(self, a):
+            self.a = a
 
-    def test_apply_to_logging(self):
-        """correctly creates log file"""
-        dstore = io_app.get_data_store("data", suffix="fasta", limit=3)
-        with TemporaryDirectory(dir=".") as dirname:
-            reader = io_app.load_aligned(format="fasta", moltype="dna")
-            min_length = sample_app.min_length(10)
-            outpath = os.path.join(os.getcwd(), dirname, "delme.tinydb")
-            writer = io_app.write_db(outpath)
-            process = reader + min_length + writer
-            r = process.apply_to(dstore, show_progress=False)
-            # always creates a log
-            self.assertEqual(len(process.data_store.logs), 1)
-            process.data_store.close()
+        def main(self, val: int) -> int:
+            return val
+
+    class app_dummyclass_2:
+        def __init__(self, b):
+            self.b = b
+
+        def main(self, val: int) -> int:
+            return val
+
+    define_app(app_dummyclass_1)
```

Comment:
why not actually decorate the classes?

Reply:
Sorry, it remains from previous format that I had defined classes outside test functions. Now I will fix them.

Reply:
I see you merged it. Next time I will change it.